File: scala/ch02/ex2.68.scala
type Weight = Int

// In contrast with the 2.6Nb exercises, we'll make Leaf a case class and not
// a subclass of TreeWithBranches. Also, we eliminated EmptyTree.
sealed abstract class Tree(val symbols: List[Symbol], val weight: Weight)
case class TreeWithBranches(
  leftBranch: Tree, rightBranch: Tree, 
  override val symbols: List[Symbol], 
  override val weight: Weight) 
    extends Tree(symbols, weight)
case class Leaf(symbol: Symbol, override val weight: Weight) 
    extends Tree(List(symbol), weight)

def makeCodeTree (left: Tree, right: Tree) = 
  TreeWithBranches(
    left, 
    right, 
    left.symbols ++ right.symbols, 
    left.weight + right.weight)
              
sealed abstract class Bit(val value: Int)
case object Zero extends Bit(0)
case object One  extends Bit(1)

class DecodeException(message: String) extends RuntimeException(message)

def decode (bits: List[Bit], tree: TreeWithBranches): List[Symbol] = {
  def decode1 (bits: List[Bit], current: TreeWithBranches): List[Symbol] = bits match {
    case Nil => Nil
    case head :: tail => chooseBranch (head, current) match {
      case Leaf(symbol, _) => symbol :: decode1(tail, tree)
      case next:TreeWithBranches => decode1(tail, next)
    }
  }
  decode1 (bits, tree)
}
  
def chooseBranch (bit: Bit, branch: TreeWithBranches) = bit match {
  case Zero => branch.leftBranch
  case One  => branch.rightBranch
}
  
def encodeSymbol (symbol: Symbol, tree: Tree) = {
  def encodeSym (subtree: Tree, result: List[Bit]): List[Bit] = subtree match {
    case Leaf(sym, _) => if (sym == symbol) result else Nil
    case twb:TreeWithBranches => 
      val leftResult  = encodeSym (twb.leftBranch,  Zero :: result)
      val rightResult = encodeSym (twb.rightBranch, One  :: result)
      if      (leftResult  != Nil) leftResult
      else if (rightResult != Nil) rightResult
      else Nil
  }
  encodeSym(tree, Nil).reverse match {
    case Nil => throw new DecodeException("Could not encode symbol")
    case answer => answer
  }
}

def encode (message: List[Symbol], tree: Tree): List[Bit] = message match {
  case Nil => Nil
  case head :: tail => encodeSymbol (head, tree) ++ encode (tail, tree)
}

val sampleTree = 
  makeCodeTree(
    Leaf ('A, 4),
    makeCodeTree(
      Leaf ('B, 2),
      makeCodeTree(
        Leaf ('D, 1),
        Leaf ('C, 1))))
        
val expectedMessage =
  List(Zero, One, One, Zero, Zero, One, Zero, 
       One, Zero, One, One, One, Zero)

import org.scalatest._ 
import org.scalatest.matchers._

object huffmanEncodeSpec extends Spec with ShouldMatchers {

  describe ("encode") {
    it ("should return the sequence of bits for a sequence of letters") {
      encode (List('A), sampleTree) should equal (List(Zero))
      encode (List('B), sampleTree) should equal (List(One, Zero))
      encode (List('C), sampleTree) should equal (List(One, One, One))
      encode (List('D), sampleTree) should equal (List(One, One, Zero))
      evaluating { 
        encode (List('E), sampleTree) 
      } should produce [DecodeException]
      encode (List('A, 'D, 'A, 'B, 'B, 'C, 'A), sampleTree) should equal (expectedMessage)
    }
  }
}
huffmanEncodeSpec execute

File: clojure/ch02/ex2.68.clj
(ns sicp.ch02 (:use clojure.test))

(defn make-leaf [symbol weight]
  ['leaf symbol weight])
(defn leaf? [x]
  (= (first x) 'leaf))
(defn symbol-leaf [x] (nth x 1))
(defn weight-leaf [x] (nth x 2))

(defn symbols [tree]
  (if (leaf? tree)
      [(symbol-leaf tree)]
      (nth tree 2)))

(defn weight [tree]
  (if (leaf? tree)
      (weight-leaf tree)
      (nth tree 3)))
      
(defn make-code-tree [left right]
  [left
   right
   (into (symbols left) (symbols right))
   (+ (weight left) (weight right))])
        
(defn left-branch [tree] (first tree))
(defn right-branch [tree] (nth tree 1))

(defn choose-branch [bit branch]
    (cond (= bit 0) (left-branch branch)
          (= bit 1) (right-branch branch)
          :else (throw (Exception. "bad bit -- choose-branch" bit))))
          
(defn decode [bits tree]
  (defn decode-1 [bits current-branch]
    (if (empty? bits)
      []
      (let [next-branch (choose-branch (first bits) current-branch)]
        (if (leaf? next-branch)
            (cons (symbol-leaf next-branch)
                  (decode-1 (rest bits) tree))
            (decode-1 (rest bits) next-branch)))))
  (decode-1 bits tree))
  
(defn encode-symbol [symbol tree]
  (defn encode-sym [subtree result]
    (if (leaf? subtree)
        (if (= symbol (symbol-leaf subtree)) result [])
        (let [left-result  (encode-sym (left-branch subtree) (conj result 0))
              right-result (encode-sym (right-branch subtree) (conj result 1))]
          (cond (not (empty? left-result)) left-result
                (not (empty? right-result)) right-result
                :else []))))
  (let [answer (encode-sym tree [])]
    (if (empty? answer)
        (throw (Exception. "Could not encode symbol"))
        answer)))

(defn encode [message tree]
  (if (empty? message)
      []
      (into (encode-symbol (first message) tree)
            (encode (rest message) tree))))

(def sample-tree
  (make-code-tree 
    (make-leaf 'A 4)
    (make-code-tree
      (make-leaf 'B 2)
      (make-code-tree 
        (make-leaf 'D 1)
        (make-leaf 'C 1)))))
        
(def expected-message [0 1 1 0 0 1 0 1 0 1 1 1 0])

(deftest test-make-code-tree
  (is (= (make-code-tree (make-leaf 'D 1) (make-leaf 'C 1)) 
         '[[leaf D 1] [leaf C 1] [D C] 2])))
  
(deftest test-encode
  (is (= (encode '[A] sample-tree) [0]))
  (is (= (encode '[B] sample-tree) [1 0]))
  (is (= (encode '[C] sample-tree) [1 1 1]))
  (is (= (encode '[D] sample-tree) [1 1 0]))
  ; (is (= (encode '[E] sample-tree) [1 1 0]))
  (is (= (encode '[A D A B B C A] sample-tree) expected-message)))
  
(run-tests)

